cpp/src/string/nvcategory_util.cpp

```diff
@@ -0,0 +1,130 @@
+
+#include "nvcategory_util.hpp"
+
+
+#include <nvstrings/NVCategory.h>
+#include <nvstrings/NVStrings.h>
+#include "rmm/rmm.h"
+
+#include "utilities/error_utils.hpp"
+#include "utilities/nvtx/nvtx_utils.h"
+
+
+gdf_error nvcategory_gather(gdf_column * column, NVCategory * nv_category){
+
+
+  GDF_REQUIRE(nv_category != nullptr,GDF_INVALID_API_CALL );
+  GDF_REQUIRE(column->dtype == GDF_STRING_CATEGORY,GDF_UNSUPPORTED_DTYPE);
+
+  NVCategory * new_category = nv_category->gather(static_cast<nv_category_index_type *>(column->data),
+                                                          column->size,
+                                                          DEVICE_ALLOCATED);
+  new_category->get_values(static_cast<nv_category_index_type *>(column->data),
+                           DEVICE_ALLOCATED);
+
+  //Python handles freeing the original column->dtype_info.category so we don't need to
+  column->dtype_info.category = new_category;
+
+
+  return GDF_SUCCESS;
+}
+
+gdf_error validate_categories(gdf_column * input_columns[], int num_columns, gdf_size_type & total_count){
+  total_count = 0;
+  for (int i = 0; i < num_columns; ++i) {
+    gdf_column* current_column = input_columns[i];
+    GDF_REQUIRE(current_column != nullptr,GDF_DATASET_EMPTY);
+    GDF_REQUIRE(current_column->data != nullptr,GDF_DATASET_EMPTY);
+    GDF_REQUIRE(current_column->dtype == GDF_STRING_CATEGORY,GDF_UNSUPPORTED_DTYPE);
+
+    total_count += input_columns[i]->size;
+  }
+  return GDF_SUCCESS;
+}
+
+NVCategory * combine_column_categories(gdf_column * input_columns[],int num_columns){
```

Comment:
Put any helper functions not exposed in the header into an anonymous namespace to make it clear they aren't external functions.